pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,565 @@
+	/// Etherium address, for cases where we would have a standalone Eth address
+	Eth(H160),
+}
+
+impl From<u64> for Location {
+	fn from(a: u64) -> Self {
+		Self::TestLocal(a)
+	}
+}
+
+impl From<AccountId32> for Location {
+	fn from(a: AccountId32) -> Self {
+		Self::Local(a)
+	}
+}
+
+impl From<MultiLocation> for Location {
+	fn from(ml: MultiLocation) -> Self {
+		// using hash here as mulitlocation is significantly larger than any other enum type here
+		// -- 592 bytes, vs 40 bytes for domain address (next largest)
+		Self::XCM(BlakeTwo256::hash(&ml.encode()))
+	}
+}
+
+impl From<VersionedMultiLocation> for Location {
+	fn from(vml: VersionedMultiLocation) -> Self {
+		// using hash here as mulitlocation is significantly larger than any other enum type here
+		// -- 592 bytes, vs 40 bytes for domain address (next largest)
+		Self::XCM(BlakeTwo256::hash(&vml.encode()))
+	}
+}
+
+impl From<DomainAddress> for Location {
+	fn from(da: DomainAddress) -> Self {
+		Self::Address(da)
+	}
+}
+
+impl From<H160> for Location {
+	fn from(eth: H160) -> Self {
+		Self::Eth(eth)
+	}
+}
+
+/// Trait to determine whether a sending account and currency have a restriction,
+/// and if so is there an allowance for the reciever location.
+pub trait TransferAllowance<AccountId> {
+	type CurrencyId;
+	/// Determines whether the `send` account is allowed to make a transfer to the  `recieve` loocation with `currency` type currency.
+	/// Returns result wrapped bool for whether allowance is allowed.
+	fn allowance(
+		send: AccountId,
+		recieve: Location,
+		currency: Self::CurrencyId,
+	) -> Result<bool, DispatchError>;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use core::fmt::Debug;
+
+	use frame_support::{
+		pallet_prelude::{DispatchResult, OptionQuery, StorageDoubleMap, StorageNMap, *},
+		traits::{tokens::AssetId, Currency, ReservableCurrency},
+		transactional, Twox64Concat,
+	};
+	use frame_system::pallet_prelude::{OriginFor, *};
+	use sp_runtime::{traits::AtLeast32BitUnsigned, Saturating};
+
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Id type of Currency transfer restrictions will be checked for
+		type CurrencyId: AssetId
+			+ Parameter
+			+ Debug
+			+ Member
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Ord
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		/// Deposit Balance to reserve/release
+		#[pallet::constant]
+		type Deposit: Get<DepositBalanceOf<Self>>;
+
+		/// Currency for Reserve/Unreserve with allowlist adding/removal,
+		/// given that the allowlist will be in storage
+		type ReserveCurrency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+	}
+
+	pub type BlockNumberOf<T> = <T as frame_system::Config>::BlockNumber;
+	pub type CurrencyIdOf<T> = <T as Config>::CurrencyId;
```

Comment:
Is this new type really needed?

Reply:
Updated to use T::CurrencyId -- same as AccountId comment updated